File: packages/node/otlp-stdout-span-exporter/examples/simple-stdout-hello.ts
/**
 * Simple example showing basic usage of OTLPStdoutSpanExporter.
 * This example demonstrates:
 * 1. Basic tracer setup
 * 2. Creating a parent span
 * 3. Creating a nested child span
 * 4. Proper cleanup with force flush
 */

import { trace, Span } from '@opentelemetry/api';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { LogLevel, OTLPStdoutSpanExporter, OutputType } from '../src';

function initTracer(): NodeTracerProvider {
  // Create a tracer provider
  const provider = new NodeTracerProvider(
    {
      // Register the exporter with the provider
      spanProcessors: [
        new BatchSpanProcessor(new OTLPStdoutSpanExporter(
          {
            gzipLevel: 9,
            logLevel: LogLevel.Info,
            outputType: OutputType.Stdout,
          }
        ))
      ]
    }
  );

  // Set as global default tracer provider
  provider.register();

  return provider;
}

async function main(): Promise<void> {
  // Initialize tracing
  const provider = initTracer();
  const tracer = trace.getTracer('example/simple');

  // Create a parent span using the correct pattern
  await tracer.startActiveSpan('parent-operation', async (parentSpan: Span) => {
    parentSpan.addEvent('Doing work...');

    // Create a nested child span
    await tracer.startActiveSpan('child-operation', async (childSpan: Span) => {
      childSpan.addEvent('Doing more work...');
      childSpan.end();
    });

    parentSpan.end();
  });

  // Ensure all spans are exported
  await provider.forceFlush();

  // Shutdown the provider
  await provider.shutdown();
}

// Run the example
main().catch(console.error);

File: packages/rust/otlp-stdout-span-exporter/examples/simple-stdout-hello.rs
use opentelemetry::global;
use opentelemetry::trace::{get_active_span, Tracer};
use opentelemetry::KeyValue;
use opentelemetry_sdk::trace::SdkTracerProvider;
use otlp_stdout_span_exporter::{LogLevel, OtlpStdoutSpanExporter};
use std::collections::HashMap;

fn init_tracer() -> SdkTracerProvider {
    let mut headers: HashMap<String, String> = HashMap::new();
    headers.insert("test".to_string(), "test".to_string());

    // Create exporter with the Debug log level and named pipe output
    // You can also use environment variables:
    // OTLP_STDOUT_SPAN_EXPORTER_LOG_LEVEL=debug
    // OTLP_STDOUT_SPAN_EXPORTER_OUTPUT_TYPE=pipe
    let exporter = OtlpStdoutSpanExporter::builder()
        .headers(headers)
        .level(LogLevel::Debug)
        .pipe(true) // Will write to /tmp/otlp-stdout-span-exporter.pipe
        .build();
    let provider = SdkTracerProvider::builder()
        .with_batch_exporter(exporter)
        .build();

    global::set_tracer_provider(provider.clone());
    provider
}

#[tokio::main]
async fn main() {
    eprintln!("Writing spans to /tmp/otlp-stdout-span-exporter.pipe with DEBUG level");
    eprintln!("Note: Make sure the named pipe exists (create with `mkfifo /tmp/otlp-stdout-span-exporter.pipe`)");

    let provider = init_tracer();
    let tracer = global::tracer("example/simple");
    tracer.in_span("parent-operation", |_cx| {
        get_active_span(|span| {
            span.add_event("Doing work".to_string(), vec![KeyValue::new("work", true)]);
        });

        // Create nested spans
        tracer.in_span("child-operation", |_cx| {
            get_active_span(|span| {
                span.add_event(
                    "Not doing work".to_string(),
                    vec![KeyValue::new("work", false)],
                );
            });
        });
    });

    if let Err(err) = provider.force_flush() {
        eprintln!("Error flushing provider: {:?}", err);
    }

    eprintln!("Spans have been written to /tmp/otlp-stdout-span-exporter.pipe");
}
